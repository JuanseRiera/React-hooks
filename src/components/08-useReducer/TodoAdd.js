import React from "react";
import PropTypes from "prop-types";
import { useForm } from "../../Hooks/useForm";

export const TodoAdd = ({ guardarTarea }) => {
  const { formState, changeInput, reset } = useForm({
    todoInput: "",
  });

  const submit = (e) => {
    e.preventDefault();
    guardarTarea({
      id: Math.round(Math.random() * 100000),
      todo: formState.todoInput,
      done: false,
    });
    reset();
  };
  return (
    <>
      <form onSubmit={submit} className="text-end">
        <input
          type="text"
          className="form-control"
          placeholder="Ingrese su tarea por hacer..."
          name="todoInput"
          autoComplete="off"
          onChange={changeInput}
          value={formState.todoInput}
        />
      </form>
    </>
  );
};

TodoAdd.propTypes = {
  guardarTarea: PropTypes.func.isRequired,
};
