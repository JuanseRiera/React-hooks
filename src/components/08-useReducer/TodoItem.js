import React from "react";
import PropTypes from "prop-types";

export const TodoItem = ({ item, toggle, eliminar }) => {
  const { id, todo, done } = JSON.parse(item);
  return (
    <>
      <div className="card mb-3">
        <div className="card-body row">
          <div className="col-12 col-sm-8 my-auto">
            <h6
              className={done ? "text-decoration-line-through" : null}
              onClick={() => {
                toggle(id);
              }}
            >
              {todo}
            </h6>
          </div>
          <div className="col-12 col-sm-4 text-end">
            <button
              className="btn btn-danger ms-2"
              onClick={() => {
                eliminar(id);
              }}
            >
              Borrar
            </button>
          </div>
        </div>
      </div>
    </>
  );
};

TodoItem.propTypes = {
  item: PropTypes.string.isRequired,
  toggle: PropTypes.func.isRequired,
  eliminar: PropTypes.func.isRequired,
};
