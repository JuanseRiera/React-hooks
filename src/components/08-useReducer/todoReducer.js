export const todoReducer = (state, action) => {
  switch (action.type) {
    case "agregar":
      return [...state, action.payload];

    case "toggle":
      return state.map((todo) => {
        if (todo.id === action.payload) {
          return {
            ...todo,
            done: !todo.done,
          };
        } else {
          return todo;
        }
      });

    case "eliminar":
      return state.filter((x) => x.id !== action.payload);

    default:
      return state;
  }
};
