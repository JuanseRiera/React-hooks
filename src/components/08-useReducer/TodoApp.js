import React, { useEffect, useReducer } from "react";
import { TodoItem } from "./TodoItem";
import { todoReducer } from "./todoReducer";
import "./reducer.css";
import { TodoAdd } from "./TodoAdd";

export const TodoApp = () => {
  const init = () => {
    return JSON.parse(localStorage.getItem("tareas")) || [];
  };

  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const guardarTarea = (todo) => {
    let actionAgregar = {
      type: "agregar",
      payload: todo,
    };
    dispatch(actionAgregar);
  };

  const toggle = (id) => {
    let actiontoggle = {
      type: "toggle",
      payload: id,
    };
    dispatch(actiontoggle);
  };

  const eliminar = (id) => {
    let actionEliminar = {
      type: "eliminar",
      payload: id,
    };
    dispatch(actionEliminar);
  };

  useEffect(() => {
    localStorage.setItem("tareas", JSON.stringify(todos));
  }, [todos]);

  return (
    <>
      <div className="row">
        <div className="col-12 col-md-4 my-auto">
          <h1>TodoApp</h1>
        </div>
        <div className="col-12 col-md-8 my-auto">
          <TodoAdd guardarTarea={guardarTarea} />
        </div>
      </div>

      <hr />

      <div className="row">
        <div className="col-12 p-3">
          {todos.map((todo) => {
            return (
              <TodoItem
                key={todo.id}
                item={JSON.stringify(todo)}
                toggle={toggle}
                eliminar={eliminar}
              />
            );
          })}
        </div>
      </div>
    </>
  );
};
