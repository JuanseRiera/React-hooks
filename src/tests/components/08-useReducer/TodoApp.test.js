import { act } from "@testing-library/react";
import { shallow, mount } from "enzyme";
import { TodoApp } from "../../../components/08-useReducer/TodoApp";
import { demoTodo } from "../../fixtures/demoTodos";

describe("Pruebas para todoApp", () => {
  const wrapper = shallow(<TodoApp />);
  Storage.prototype.setItem = jest.fn();
  test("Se deberia cargar el documento correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe cargarse dos elementos y se debe llamar dos veces a setItem del localstorage", () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find("TodoAdd").prop("guardarTarea")(demoTodo[0]);
      wrapper.find("TodoAdd").prop("guardarTarea")(demoTodo[1]);
    });

    console.log(wrapper.find("TodoItem"));
    expect(localStorage.setItem).toBeCalledTimes(2);
  });
});
