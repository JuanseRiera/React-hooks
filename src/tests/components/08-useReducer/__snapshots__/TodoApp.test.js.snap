// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para todoApp Se deberia cargar el documento correctamente 1`] = `
<Fragment>
  <div
    className="row"
  >
    <div
      className="col-12 col-md-4 my-auto"
    >
      <h1>
        TodoApp
      </h1>
    </div>
    <div
      className="col-12 col-md-8 my-auto"
    >
      <TodoAdd
        guardarTarea={[Function]}
      />
    </div>
  </div>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-12 p-3"
    />
  </div>
</Fragment>
`;
